api/proto/teleport/userpreferences/v1/pinned_resources.proto

```diff
@@ -0,0 +1,32 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.userpreferences.v1;
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/userpreferences/v1;userpreferencesv1";
+
+// ClusterPinnedResources is a list of resource ids pinned per cluster
+message ClusterPinnedResources {
+  // resource_ids is a list of resource ids
+  repeated string resource_ids = 1;
+}
+
+// PinnedResourcesUserPreferences is a collection of resource IDs that will be
+// displayed in the user's pinned resources tab in the Web UI
+message PinnedResourcesUserPreferences {
+  // pinned_resources is a map of resource IDs pinned per cluster
+  map<string, ClusterPinnedResources> pinned_resources = 1;
```

Comment:
> Should pinned resources be stored per cluster rather than having them all in the root cluster?

I'm not sure yet. It seems reasonable enough to allow it but I haven't done any super thinking on it yet.

> Is there an upper bound to the number of pinned resources? Is it per cluster or total? 

There isn't an upperbound programmed in yet, but we'd want to stay under the max size of an item, lets say in dynamodb, which is 400kb. if we have a node id like "123123-123123123-123123123/node", well over 10k could fit and still have plenty of room for more (and other user preferences).

I don't think 10k pinned resources across all clusters is a realistic outcome in terms of what people actually pin, but we should put some safeguards up regardless. and also it comes into play with my next point. 

> How do pinned resources get cleaned up? Is it manual? What happens if a leaf cluster is removed entirely?

This will require some extra thinking. My first inclination is that when a pinned resources request is made, we can track how many resources "couldn't be found" from their pinned list, and then just remove them. But I don't think thats a good place to do it as I wouldn't want to add extra latency to the request. there should be some way to clean them up however, because if that was left to grow unbounded then we could run into the issue above.

> What will be displayed to a user if a pinned resource is removed? 

If they pinned something that gets removed or they lose access to it, it just wouldn't be in shown. a "pinned resource" is really just an extra filter with specific IDs. you can do the same with an advanced query, so the same UX if they opened the normal tab and a resource was gone. it just wouldn't be there. 

> Is the resource id enough information for a user to discern which resource they pinned?

This isn't super relevant yet since they won't see something that doesn't exist. However, I could envision a future where there is a 'user preferences' page that could show the IDs they have pinned (existing and not) and be able to edit it there. Then maybe ID wouldn't be enough, but ID + if it exists or not (with info from existing resources surrounding it) might be enough. That bridge is very far down the road tho.


To summarize, the main points that should probably be solved now-ish or at least documented are the cleanup phase and the max size. I will try to get something written up about this

Reply:
Based on the above, I'd lean toward storing the pins per-cluster rather than storing all pins in root as is currently done.  If we store pins per cluster, doing automatic cleanup as a periodic won't be too difficult.  If, on the other hand, we do all pinning in the root that makes things _very_ complicated.  Cross-cluster communication is non-trivial.

Also note that a resource disappearing temporarily doesn't necessarily mean that it's gone.  An automatic cleanup routine should probably track timestamps and only clean up resource that remain missing for some minimum amount of time + number of checks.

IMO wether or not this PR includes a cleanup mechanism, we should have a reasonably clear idea of how the cleanup mechanism will function prior to merging it.

Reply:
I don't think it's a bad idea to put an upper limit on the number of pinned resources. Pinning resources becomes less helpful the more you pin, so I'd expect people won't pin too many. Maybe we set a hard cap to something like 300 or 500?

As far as cleanup goes, I think for now it's okay if your preferences reference a resource that no longer exists - they just won't show up in the listing.

We can work with @roraback to get some designs for how you might remove them.